File: cpp/medium/CanPartition.cpp
// 416. 分割等和子集

#include <algorithm>
#include <vector>
using namespace std;

bool canPartition(vector<int> &nums) {
  if (nums.size() <= 1)
    return false;
  int count = 0;
  for (int num : nums) {
    count += num;
  }
  // 如果总和为奇数则必然不可能拆分为两个相等的子集
  if (count % 2)
    return false;
  count = count / 2;
  vector<vector<int>> dp(nums.size(), vector<int>(count + 1, 0));
  // 初始化
  for (int i = 0; i < nums.size(); i++) {
    dp[i][0] = 0;
  }
  for (int i = 0; i <= count; i++) {
    if (i >= nums[0])
      dp[0][i] = nums[0];
  }
  for (int i = 1; i < nums.size(); i++) {
    for (int k = 1; k <= count; k++) {
      if (nums[i] > k)
        dp[i][k] = dp[i - 1][k];
      else
        dp[i][k] = max(dp[i - 1][k], dp[i - 1][k - nums[i]] + nums[i]);
    }
  }
  for (int i = 0; i < nums.size(); i++) {
    if (dp[i][count] == count)
      return true;
  }
  return false;
}

bool canPartition2(vector<int> &nums) {
  if (nums.size() <= 1) {
    return false;
  }
  int sum = 0;
  for (int i = 0; i < nums.size(); i++)
    sum += nums[i];
  if (sum % 2)
    return false;
  int target = sum / 2;
  vector<int> dp(target + 1, 0);

  for (int i = 0; i < nums.size(); i++) {
    for (int k = target; k >= nums[i]; k--) {
      dp[k] = max(dp[k], dp[k - nums[i]] + nums[i]);
    }
  }
  if (dp[target] == target)
    return true;
  return false;
}

int main() { vector<int> nums = {1, 2, 5}; }

File: go/easy/canPartition.go
package main

import "fmt"

func canPartition(nums []int) bool {
	sum := 0
	for _, v := range nums {
		sum += v
	}
	if sum%2 == 1 {
		return false
	}
	dp := make([][]int, sum+1)
	for i := range dp {
		dp[i] = make([]int, len(nums)+1)
	}
	for i := 1; i < len(dp); i++ {
		for k := 1; k < len(dp[0]); k++ {
			if nums[k-1] <= i {
				dp[i][k] = max(dp[i-nums[k-1]][k-1]+nums[k-1], dp[i][k-1])
			} else {
				dp[i][k] = dp[i][k-1]
			}
		}
	}
	for i := range dp {
		for k := range dp[0] {
			fmt.Print(dp[i][k], " ")
		}
		fmt.Println()
	}
	return dp[len(dp)-1][len(dp[0])-1] == sum/2
}

func main() {
	nums := []int{1, 5, 11, 5}
	canPartition(nums)
}
